### PR TITLE
Dpl498 location report add retention instructions

### LINKED ISSUE #3674: DPL-498 Report on retention instructions
**User story**
As a Sample Management user, I would like the plate location report to include any 'Retention instructions' recorded against the labware, so that I can find out what to do with old plates. The other information in the location report will also be useful in this context for cross-referencing.

**Who are the primary contacts for this story**
Danni W, Siobhan A-G, Katy

**Who is the nominated tester for UAT**
Danni W, Siobhan A-G

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Plate location report includes new column, that contains 'Retention instructions', if there are any against the labware

**Dependencies**
This story is blocked by the following dependencies:
https://github.com/sanger/sequencescape/issues/3673 - this is where the 'Retention instructions' will be added.

**Additional information**
Location report can be found at https://training.sequencescape.psd.sanger.ac.uk/location_reports

### PR DESCRIPTION
**Pull request by andrewsparkes**
Closes #3674

### CODE REVIEW

File: app/sample_manifest_excel/sample_manifest_excel/upload/processor/plate.rb [hunk 1]
```diff
@@ -60,6 +62,59 @@ def find_plate_id_for_sample_id(sample_id)
...
+        def non_matching_retention_instructions_for_plates
+          return nil, nil unless upload.respond_to?(:rows)
+
+          upload
+            .rows
+            .each_with_object({}) do |row, plate_retentions|
+              # ignore empty rows and skip if the retention column is not present
+              if row.columns.blank? || row.data.blank? || row.columns.extract(['retention_instruction']).count.zero?
+                next
+              end
```
Comment (sdjmchattie) [inline app/sample_manifest_excel/sample_manifest_excel/upload/processor/plate.rb:86]: You could avoid having this `if` check with a `next` operation if you add a `.reject` operation before your `.each_with_object` operation.

File: app/sample_manifest_excel/sample_manifest_excel/upload/processor/plate.rb [hunk 2]
```diff
@@ -60,6 +62,59 @@ def find_plate_id_for_sample_id(sample_id)
...
+
+              plate_barcode = row.value('sanger_plate_id')
+              sample_id = row.value('sanger_sample_id')
+
+              # ignore rows where primary sample fields have not been filled in
+              next unless plate_barcode.present? && sample_id.present?
```
Comment (sdjmchattie) [inline app/sample_manifest_excel/sample_manifest_excel/upload/processor/plate.rb:92]: You could even add this to your `.reject` operation

**Review by sdjmchattie — APPROVED** [review]
One minor suggestion

File: app/models/location_report.rb [hunk 3]
```diff
@@ -148,6 +148,15 @@ def generate_plate_cols_for_row(cur_plate)
     cols << cur_plate.created_at.strftime('%Y-%m-%d %H:%M:%S')
     cols << cur_plate.storage_location
     cols << cur_plate.storage_location_service
+    cols << (get_retention_instructions(cur_plate) || 'Unknown')
+  end
+
+  def get_retention_instructions(cur_plate)
```
[the remark this answers is not in the document]
Comment (andrewsparkes) [reply app/models/location_report.rb:154]: is this suggesting the method should go in the plate model?

[the remark this answers is not in the document]
Comment (sdjmchattie) [reply app/models/location_report.rb:154]: It's indicating that you don't refer to `self.` or any other instance methods or attributes.  It operates entirely on the thing you passed in `cur_plate` and then returns a value.  So it ought to be in a separate helper file, or perhaps you could indeed put it in the `Plate` model as that's what it operates on.  You could create an extension for the Plate class by "reopening" it and defining this method on it if you'd prefer to put it in a different file.  http://juixe.com/techknow/index.php/2007/01/17/reopening-ruby-classes-2

Not sure if that's preferable here or not.

**Comment by andrewsparkes** [comment]
merged into 499 so not needed

**Pull request closed without merging**